exercicios/para-casa/ETL_JessicaBom.py

```diff
@@ -0,0 +1,51 @@
+import pandas as pd
+
+df = pd.read_csv("../../material/mais_ouvidas_2024.csv")
+
+print(df.head())
+print(df.dtypes)
+
+
+#Aqui criamos uma variável que armazena em uma lista (nome_colunas) as colunas existentes em um csv, sem a necessidade de copiar e colar os valores 
+nome_colunas = df.columns.tolist()
+
+
+#Para converter as colunas numéricas em formato de números não usei replace e astype. Usei direto o to_numeric e usei o replace para retirar as vírgulas em um for para fazer uma série de transformações. (1) analisa se a coluna é do tipo ocject, se sim (2) aplica o to_numeric, checando se aquele valor pode ser transformado em float ou int. Contudo, como haviam vírgulas em alguns números essa função não completava, por isso, apliquei o str.replace para trocar todas as vírgulas que ele encontrasse por vazio (""). Os erros foram ignorados, mas os espaços vazios foram substituídos por NaN. Preferi não filtrar com o fillna porque mudava o tipo da coluna como um todo para object quando havia um espaço vazio. 
+for coluna in nome_colunas:
+    if df[coluna].dtype == "object":
+        df[coluna] = pd.to_numeric(df[coluna].str.replace(",", ""), errors = "ignore")
```

Comment:
vai tentar converter todas as colunas que tem o tipo `object` isso inclui as colunas que DEVEM ser object e que não podem ser convertidas usando `astype`. Por isso usou to_numeric?

qual a diferença entre elas?

Reply:
quando fiz o código, achei mais prático usar o to_numeric, porque ele identifica meio que "automaticamente" o que pode ser numérico e já coloca como float ou como int pela própria função. Da primeira vez usei com o coerce e vi que ele transformava o que de fato tinha que ser object em NaN. Depois refiz utilizando o "ignore". Meu pensamento foi: se o to_numeric identifica o que é numérico e faz a conversão, eu faço um for passando por todas as coluna da variável "nome_colunas" e se aquilo não puder ser um número, o to_numeric vai ignorar com o "ignore" e aquilo permanece como string ou o outro formato.

Reply:
Esse é uma ótima linha de pensamento!

Trago pra você outra análise então:

Se o ˋto_numericˋ  detecta "automaticamente" se o valor pode ou não ser convertido, porquê os erros aconteceram? Quando tentamos converter com os astype se não for possível converter também recebemos erro. Em quê eles são diferentes?

Reply:
Essa pergunta aqui pegou professora. Fui olhar a documentação pra entender a diferença e depois recorri ao chatGPT porque não estava muito claro. Mas lá vai o que eu entendi.
Os erros usando o "to_numeric" aconteceram porque ele encontrou no DataFrame elementos que não podiam ser convertidos para números naquelas colunas que eu especifiquei.

Pelo que vi aqui, em tese não teria muita diferença, porque por padrão o retorno do "to_numeric" e do "astype" pra erros é o raise, ou seja, quando eles encontram um elemento que não pode ser convertido para o tipo desejado eles retornariam um erro imediato, um "ValueError"(acho).

Mas, assim como o "to_numeric", também podemos configurar o "as_type" para ignorar os erros. Pelo que vi na documentação ele tem por padrão errors=raise, mas também pode receber errors=ignore.
Provavelmente, nesse caso, o problema maior aqui é que o to_numeric pode converter os valores para NaN se configurarmos o errors=coerce, enquanto o astype não pode receber isso. Então - se for isso e se eu não embananei tudo - usar o astype para a conversão acaba sendo duplamente mais seguro. Seria isso?